microprofile/connectors/kafka/src/main/java/io/helidon/microprofile/connectors/kafka/KafkaPublisher.java

```diff
@@ -0,0 +1,311 @@
+        this.pollTimeout = pollTimeout;
+        this.maxEvents = maxEvents;
+        this.autoCommit = autoCommit;
+        this.ackTimeout = ackTimeout;
+        this.limitNoAck = limitNoAck;
+    }
+
+    /**
+     * Starts to consume events from Kafka to send them downstream till
+     * {@link io.helidon.microprofile.connectors.kafka.KafkaPublisher#close()} is invoked.
+     * This execution runs in one thread that is triggered by the scheduler.
+     */
+    private void execute() {
+        kafkaConsumer.subscribe(topics, partitionsAssignedLatch);
+        // This thread reads from Kafka topics and push in kafkaBufferedEvents
+        scheduler.scheduleAtFixedRate(() -> {
+            try {
+                // Need to lock to avoid onClose() is executed meanwhile task is running
+                taskLock.lock();
+                if (!scheduler.isShutdown() && !emiter.isTerminated()) {
+                    int currentNoAck = currentNoAck();
+                    if (currentNoAck < limitNoAck) {
+                        if (backPressureBuffer.isEmpty()) {
+                            try {
+                                kafkaConsumer.poll(Duration.ofMillis(pollTimeout)).forEach(backPressureBuffer::add);
+                            } catch (WakeupException e) {
+                                LOGGER.fine(() -> "It was requested to stop polling from channel");
+                            }
+                        } else {
+                            long totalToEmit = requests.get();
+                            // Avoid index out bound exceptions
+                            long eventsToEmit = Math.min(totalToEmit, backPressureBuffer.size());
+                            for (long i = 0; i < eventsToEmit; i++) {
+                                if (maxEvents == 0) {
+                                    emiter.complete();
+                                    break;
+                                }
+                                ConsumerRecord<K, V> cr = backPressureBuffer.poll();
+                                KafkaMessage<K, V> kafkaMessage = new KafkaMessage<>(cr, autoCommit, ackTimeout);
+                                if (!autoCommit) {
+                                    TopicPartition key = new TopicPartition(kafkaMessage.getPayload().topic(),
+                                            kafkaMessage.getPayload().partition());
+                                    pendingCommits.computeIfAbsent(key, k -> new LinkedList<>()).add(kafkaMessage);
+                                }
+                                // Note that next execution will reach the user code inside @Incoming method.
+                                // By spec, onNext MUST NOT block the Publisher, otherwise it will make problems.
+                                runInNewContext(() ->  emiter.emit(kafkaMessage));
+                                requests.decrementAndGet();
+                                maxEvents--;
```

Comment:
Spotbugs is "picky", `maxEvents` doesn't have to be volatile at all. Kafka publisher should be unbounded in runtime so `maxEvents` shoudn't be needed. It can be tested like this:
```java
public Publisher<KafkaMessage<String, Long>> createPublisher(long elements) {
...
        return ReactiveStreams.fromPublisher(
                KafkaPublisher.build(Executors.newScheduledThreadPool(2), 
                        kafkaConsumer, 
                        Arrays.asList(TEST_TOPIC_1), 
                        1L, 
                        POLL_TIMEOUT, 
                        true))
                .limit(elements);
```